File: src/main/java/core/procedures/io/CloseInputPort.java
package core.procedures.io;

import core.procedures.AFn;
import core.procedures.FnArgsBuilder;
import core.scm.InputPort;
import core.scm.Void;

import java.io.IOException;

public final class CloseInputPort extends AFn {

  public CloseInputPort() {
    super(new FnArgsBuilder().min(1).max(1).mandatory(new Class[]{InputPort.class}).build());
  }

  @Override
  public String getName() {
    return "close-input-port";
  }

  @Override
  public Void apply1(Object arg) {
    try {
      ((InputPort)arg).close();
    } catch (IOException e) {
      e.printStackTrace();
    }
    return Void.VOID;
  }
}

File: src/main/java/core/procedures/io/CloseInputPort.kt
package core.procedures.io

import core.procedures.AFn
import core.procedures.Arity.Exactly
import core.scm.InputPort

class CloseInputPort : AFn<InputPort?, Unit>(name = "close-input-port", arity = Exactly(1),
                                             mandatoryArgsTypes = arrayOf(InputPort::class.java)) {

    override operator fun invoke(arg: InputPort?) = arg!!.close()
}
